### PR TITLE
oscillators: sa5x: introduce SA5x MAC oscillator

### PR DESCRIPTION
**Pull request by vvfedorenko**
add new protocol for oscillator type SA5x

### CODE REVIEW

**Comment by leoleovich** [comment]
image

File: src/oscillators/sa5x_oscillator.c [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+};
+
+struct sa5x_attributes {
+
+    uint32_t alarms;         // Alarams bits
+
+    int32_t  phaseoffset;    // Most recent phase offset
+    int32_t  lastcorrection; // Most recent freq correction
+                             // due to disciplining
+	int32_t  temperature;    // Temperature (mDegC)
```
Comment (pmazzini) [inline src/oscillators/sa5x_oscillator.c:49]: Indentation looks wrong

File: src/oscillators/sa5x_oscillator.c [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+}
+
+static int sa5x_oscillator_get_attributes(struct oscillator *oscillator, struct sa5x_attributes *a,
+                                          unsigned int attributes_mask)
+{
+	struct sa5x_oscillator *sa5x;
+	sa5x = container_of(oscillator, struct sa5x_oscillator, oscillator);
+	int err;
+
+    if (attributes_mask & (ATTR_CTRL|ATTR_STATUS) && !a) {
```
Comment (CharlesParent) [inline src/oscillators/sa5x_oscillator.c:190]: Indentation seems odd here as well

**Review by CharlesParent — CHANGES_REQUESTED** [review]

File: src/oscillators/sa5x_oscillator.c (hunk 1 above)
Comment (vvfedorenko) [reply src/oscillators/sa5x_oscillator.c]: fixed, thanks!

File: src/oscillators/sa5x_oscillator.c (hunk 2 above)
Comment (vvfedorenko) [reply src/oscillators/sa5x_oscillator.c]: fixed, thanks!

File: src/oscillators/sa5x_oscillator.c [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+		}
+
+		err = sa5x_oscillator_cmd(sa5x, CMD_GET_DIGITAL_TUNING, sizeof(CMD_GET_DIGITAL_TUNING));
+		if (err > 0) {
+			int val;
+			if (sa5x_oscillator_read_intval(&val) == 1) {
+				a->digitaltuning = val;
+			}
+			memset(answer_str, 0, err);
+		}
```
Comment (CharlesParent) [inline src/oscillators/sa5x_oscillator.c:280]: @vvfedorenko The last 8 lines are repeated multiple times, don't you think we should factorize this ?

**Review by CharlesParent — CHANGES_REQUESTED** [review]

File: src/oscillators/sa5x_oscillator.c (hunk 3 above)
Comment (vvfedorenko) [reply src/oscillators/sa5x_oscillator.c:245]: yes, Charles, you are right. Updated the diff

**Review by CharlesParent — APPROVED** [review]

**Pull request merged**